src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -0,0 +1,216 @@
+                        }
+                    }
+                    Some(&rl::Region::Static) |
+                    Some(&rl::Region::EarlyBound(_, _)) |
+                    Some(&rl::Region::LateBound(_, _)) |
+                    Some(&rl::Region::Free(_, _)) |
+                    None => {
+                        debug!("no arg found");
+                    }
+                }
+            }
+            _ => {}
+        }
+        // walk the embedded contents: e.g., if we are visiting `Vec<&Foo>`,
+        // go on to visit `&Foo`
+        intravisit::walk_ty(self, arg);
+    }
+}
+
+impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+    // This function calls the `visit_ty` method for the parameters
+    // corresponding to the anonymous regions. The `nested_visitor.found_type`
+    // contains the anonymous type.
+    pub fn find_anon_type(&self, region: Region<'tcx>, br: &ty::BoundRegion) -> Option<&hir::Ty> {
+        if self.is_suitable_anonymous_region(region).is_some() {
+            let def_id = self.is_suitable_anonymous_region(region).unwrap();
+            let node_id = self.tcx.hir.as_local_node_id(def_id).unwrap();
+            let ret_ty = self.tcx.type_of(def_id);
+            match ret_ty.sty {
+                ty::TyFnDef(_, _) => {
+                    match self.tcx.hir.get(node_id) {
+                        hir_map::NodeItem(it) => {
+                            match it.node {
+                                hir::ItemFn(ref fndecl, _, _, _, _, _) => {
+                                    fndecl
+                                        .inputs
+                                        .iter()
+                                        .filter_map(|arg| {
+                                            let mut nested_visitor = FindNestedTypeVisitor {
+                                                infcx: &self,
+                                                hir_map: &self.tcx.hir,
+                                                bound_region: *br,
+                                                found_type: None,
+                                            };
+                                            nested_visitor.visit_ty(&**arg);
+                                            if nested_visitor.found_type.is_some() {
+                                                return nested_visitor.found_type;
+                                            } else {
+                                                return None;
+                                            }
+                                        })
+                                        .next()
+                                }
+                                _ => None,
+                            }
+                        }
+                        _ => None,
+                    }
+                }
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
```

Comment:
The following should be (at least close to) correct and equivalent to the code you wrote:

```rust
fn find_anon_type(&self, region: Region<'tcx>, br: &ty::BoundRegion) -> Option<&hir::Ty> {
    if let Some(anon_reg) = self.is_suitable_anonymous_region(region) {
        let (def_id, _) = anon_reg;
        if let Some(node_id) = self.tcx.hir.as_local_node_id(def_id) {
            let ret_ty = self.tcx.type_of(def_id);
            if let ty::TyFnDef(_, _) = ret_ty.sty {
                if let hir_map::NodeItem(it) = self.tcx.hir.get(node_id) {
                    if let hir::ItemFn(ref fndecl, _, _, _, _, _) = it.node {
                        return fndecl
                            .inputs
                            .iter()
                            .filter_map(|arg| {
                                let mut nested_visitor = FindNestedTypeVisitor {
                                    infcx: &self,
                                    hir_map: &self.tcx.hir,
                                    bound_region: *br,
                                    found_type: None,
                                };
                                nested_visitor.visit_ty(&**arg);
                                if nested_visitor.found_type.is_some() {
                                  nested_visitor.found_type
                                } else {
                                  None
                                }
                            }).next();
                    }
                }
            }
        }
    }
    None
}
```

Notice the fall through for all the else conditions, with only one implicit return at the end.